src/coreclr/jit/lowerxarch.cpp

```diff
@@ -11124,6 +11124,14 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
                                     uint32_t maskSize = genTypeSize(simdBaseType);
                                     uint32_t operSize = genTypeSize(op2->AsHWIntrinsic()->GetSimdBaseType());
 
+                                    // Avx512F.Insert/ExtractVector256 accepts inputs with base type smaller that 64
+                                    // bits. it makes operSize wrong in those case.
+                                    if (op2->AsHWIntrinsic()->GetHWIntrinsicId() == NI_AVX512F_InsertVector256 ||
+                                        op2->AsHWIntrinsic()->GetHWIntrinsicId() == NI_AVX512F_ExtractVector256)
+                                    {
+                                        operSize = 8;
+                                    }
```

Comment:
This will fix the issue, but I'm not sure its entirely the "right" fix (and may miss some other similar cases).

We have a few instructions (namely those listed here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/instr.cpp#L110) where they effectively operate "bitwise" without masking but which need to consider the actual operand size when masking (or broadcast, if supported) is used.

So we probably need to ensure all of these are handled and are picking the "right" operand size (generally 4 when the input is of a "small type"). 

In the particular case of `Insert/ExtractVector256` we should probably be using `vinserti32x8` instead of `vinserti64x4` for the small types and for `int`/`uint`. Those are technically `AVX512DQ`, but the JIT relies on all of `F+BW+CD+DQ+VL` being available together, so it should be fine to just adjust them in `hwintrinsiclistxarch.h` and get the "better" codegen. Some of the other intrinsics that need special handling may be in the same boat.

Reply:
> This will fix the issue, but I'm not sure its entirely the "right" fix (and may miss some other similar cases).
> 
> We have a few instructions (namely those listed here: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/instr.cpp#L110) where they effectively operate "bitwise" without masking but which need to consider the actual operand size when masking (or broadcast, if supported) is used.

Thanks for the inputs,

At a quick glance on the instructions on the list, `and/andn/or/xor/insert/extract` may hit the same bug, `movudqx/round/broadcast` does not have EmbMask support yet. Do we consider fix them all? or leave those unsupported instructions for now (those instructions intrinsically have the support for embedded masking).

> So we probably need to ensure all of these are handled and are picking the "right" operand size (generally 4 when the input is of a "small type").
> 
> In the particular case of `Insert/ExtractVector256` we should probably be using `vinserti32x8` instead of `vinserti64x4` for the small types and for `int`/`uint`. Those are technically `AVX512DQ`, but the JIT relies on all of `F+BW+CD+DQ+VL` being available together, so it should be fine to just adjust them in `hwintrinsiclistxarch.h` and get the "better" codegen. Some of the other intrinsics that need special handling may be in the same boat. -- Alternatively we could just explicitly lower cases like `NI_AVX512F_InsertVector256` to be `NI_AVX512DQ_InsertVector256` and change the base type to `int/uint` for `byte/ubyte`, `short/ushort`, and `int/uint`. We could similarly lower cases like `NI_AVX512F_And` to normalize the base type to `int/uint`, which would likely also fix the issue

I would try to use the first option, as this is how those logical instructions (And/Or/AndN/Xor) are designed as well and based on how Avx512 ISA family is being checked, it should safe to use DQ instructions. These might still need to be handled together during lowering to normalize the base data type to `int/uint` unless it is a long type. (did some extensive testing, `And` with base type to be byte will have the same bug.)